src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -41,29 +44,35 @@ void GcsResourceManager::HandleUpdateResources(
     rpc::SendReplyCallback send_reply_callback) {
   NodeID node_id = NodeID::FromBinary(request.node_id());
   RAY_LOG(DEBUG) << "Updating resources, node id = " << node_id;
-  auto iter = cluster_resources_.find(node_id);
-  std::unordered_map<std::string, double> to_be_updated_resources;
-  for (const auto &entry : request.resources()) {
-    to_be_updated_resources.emplace(entry.first, entry.second.resource_capacity());
+  std::unordered_map<std::string, double> changed_resources;
+  for (auto &entry : request.resources()) {
+    changed_resources.emplace(entry.first, entry.second.resource_capacity());
   }
 
-  if (iter != cluster_resources_.end()) {
-    for (const auto &entry : request.resources()) {
-      (*iter->second.mutable_items())[entry.first] = entry.second;
+  auto iter = cluster_scheduling_resources_.find(node_id);
+  if (iter != cluster_scheduling_resources_.end()) {
+    // Update `cluster_scheduling_resources_`.
+    SchedulingResources &scheduling_resources = iter->second;
+    for (const auto &entry : changed_resources) {
+      scheduling_resources.UpdateResourceCapacity(entry.first, entry.second);
+    }
+
+    // Update gcs storage.
+    rpc::ResourceMap resource_map;
+    for (const auto &entry : iter->second.GetTotalResources().GetResourceMap()) {
+      (*resource_map.mutable_items())[entry.first].set_resource_capacity(entry.second);
     }
-    UpdateResourceCapacity(node_id, to_be_updated_resources);
-    auto on_done = [this, node_id, to_be_updated_resources, reply,
+    for (const auto &entry : changed_resources) {
+      (*resource_map.mutable_items())[entry.first].set_resource_capacity(entry.second);
+    }
+
+    auto on_done = [this, node_id, changed_resources, reply,
```

Comment:
It is better to make `changed_resources` as a type of shared_ptr to avoid data copy in the lambda.